File: problems/Binary Tree Maximum Path Sum/maxPathSum.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int MaxPathSum(TreeNode root) {
        var result = int.MinValue;

        getMax(root, ref result);

        return result;
    }

    private int getMax(TreeNode node, ref int res) {
        if (null == node) {
            return 0;
        }

        var leftMax = getMax(node.left, ref res);
        var rightMax = getMax(node.right, ref res);

        res = Math.Max(res, node.val + leftMax + rightMax);

        return Math.Max(0, node.val + Math.Max(leftMax, rightMax));
    }
}

File: problems/Binary Tree Maximum Path Sum/maxPathSum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxPathSum(TreeNode root) {
        dfs(root);

        return (int)result;
    }

    private long result = Integer.MIN_VALUE;

    private int dfs(TreeNode node) {
        if (null == node) {
            return 0;
        }

        int leftPathSum = dfs(node.left);
        int rightPathSum = dfs(node.right);
        int maxNodeAndOnePathSum = node.val + Math.max(leftPathSum, rightPathSum);
        int maxNodeAndTwoPathSum = Math.max(node.val, leftPathSum + node.val + rightPathSum);

        result = Math.max(result, Math.max(maxNodeAndOnePathSum, maxNodeAndTwoPathSum));

        return Math.max(node.val, maxNodeAndOnePathSum);
    }
}
